server/src/main/java/org/elasticsearch/cluster/service/MasterService.java

```diff
@@ -185,6 +195,50 @@ public String describeTasks(List<? extends BatchedTask> tasks) {
                 );
             }
 
+            public void onFailure(Exception e) {
+                try (ThreadContext.StoredContext ignore = threadContextSupplier.get()) {
+                    listener.onFailure(e);
+                } catch (Exception inner) {
+                    inner.addSuppressed(e);
+                    logger.error("exception thrown by listener notifying of failure", inner);
+                }
+            }
+
+            public void onNoLongerMaster() {
+                try (ThreadContext.StoredContext ignore = threadContextSupplier.get()) {
+                    listener.onNoLongerMaster();
+                } catch (Exception e) {
+                    logger.error("exception thrown by listener while notifying no longer master", e);
+                }
+            }
+
+            public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+                try (ThreadContext.StoredContext ignore = threadContextSupplier.get()) {
+                    listener.clusterStateProcessed(oldState, newState);
+                } catch (Exception e) {
+                    logger.error(() -> new ParameterizedMessage("""
+                        exception thrown by listener while notifying of cluster state, old cluster state:
+                        {}
+                        new cluster state:
+                        {}""", oldState, newState), e);
+                }
+            }
```

Comment:
Should we implement `ContextPreservingClusterStateTaskListener` to make thread context  related code centralized and reusable by other listeners if needed?

Reply:
My plan is that eventually we will be able to use a regular `ContextPreservingActionListener`, but I think it will be easier to get there if we keep things unwrapped like this for now.